File: frontEnd(angular)/src/app/services/student.service .ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Student } from '../models/student.models';
import { environment } from 'src/environments/environment';
import { PageStudent } from '../models/profPage.models';

@Injectable({
  providedIn: 'root',
})
export class StudentService {
  private readonly url = `${environment.backendHost}/students`;

  constructor(private http: HttpClient) {}

  // Create a new student
  createStudent(student: Student,class_id:any): Observable<Student> {
    return this.http.post<Student>(this.url+"/"+class_id, student);
  }

  // Get a student by ID
  getStudentById(id: number): Observable<Student> {
    return this.http.get<Student>(`${this.url}/${id}`);
  }

  // Update a student
  updateStudent(id: number, student: Student): Observable<Student> {
    return this.http.put<Student>(`${this.url}/${id}`, student);
  }

  // Delete a student
  deleteStudent(id: number): Observable<void> {
    return this.http.delete<void>(`${this.url}/${id}`);
  }

  // Get all students
  getAllStudents(): Observable<Student[]> {
    return this.http.get<Student[]>(this.url);
  }

  // Search students by keyword
  searchStudents(keyword: string, page: number, size: number): Observable<PageStudent> {
    const params = new HttpParams()
      .set('keyword', keyword)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageStudent>(`${this.url}/search`, { params });
  }

  // Find students by class ID
  findStudentsByClassId(classId: number): Observable<Student[]> {
    return this.http.get<Student[]>(`${this.url}/class/${classId}`);
  }

  // Search students by class and keyword
  findStudentsByClassAndKeyword(
    classId: number,
    keyword: string,
    page: number,
    size: number
  ): Observable<PageStudent> {
    const params = new HttpParams()
      .set('keyword', keyword)
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageStudent>(`${this.url}/class/${classId}/search`, { params });
  }

  public countStudent(classId: number):Observable<number> {
    return this.http.get<number>(this.url+"/class/count/" +classId  );
  }


}

File: backend(spring boot)/src/main/java/com/alialperen/timeTableGenerator/service/StudentService.java
package com.alialperen.timeTableGenerator.service;

import com.alialperen.timeTableGenerator.entity.Student;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface StudentService {
    Long countByClassId(@Param("classId") Long classId);
    Student saveStudent(Student student);

    Optional<Student> getStudentById(Long id);

    Student updateStudent(Long id, Student updatedStudent);

    void deleteStudent(Long id);

    List<Student> getAllStudents();

    Page<Student> searchStudents(String keyword, Pageable pageable);

    List<Student> findStudentsByClassId(Long classId);

    Page<Student> findStudentsByClassAndKeyword(String keyword, Long classId, Pageable pageable);

}
